flash/partition_info/partition_info.c

```diff
@@ -0,0 +1,212 @@
+
+    if (p->has_id) {
+        uint32_t id_low  = pt->table[pos++];
+        uint32_t id_high = pt->table[pos++];
+        p->partition_id = ((uint64_t)id_high << 32) | id_low;
+    } else {
+        p->partition_id = 0;
+    }
+    pt->pos = pos;
+
+    p->extra_family_id_count = (p->flags_and_permissions & PICOBIN_PARTITION_FLAGS_ACCEPTS_NUM_EXTRA_FAMILIES_BITS)
+                                   >> PICOBIN_PARTITION_FLAGS_ACCEPTS_NUM_EXTRA_FAMILIES_LSB;
+    if (p->extra_family_id_count | p->has_name) {
+        // Read variable length fields
+        uint32_t extra_family_ids_and_name[PARTITION_EXTRA_FAMILY_ID_MAX + (((PARTITION_NAME_MAX + 1) / sizeof(uint32_t)) + 1)];
+        uint32_t flags = PT_INFO_SINGLE_PARTITION | PT_INFO_PARTITION_FAMILY_IDS | PT_INFO_PARTITION_NAME;
+        int rc = rom_get_partition_table_info(extra_family_ids_and_name, sizeof(extra_family_ids_and_name),
+                                              (pt->current_partition << 24 | flags));
+        if (rc < 0) {
+            pt->status = rc;
+            return false;
+        }
+        size_t pos_ = 0;
+        uint32_t __attribute__((unused)) fields = extra_family_ids_and_name[pos_++];
+        assert(fields == flags);
+        for (size_t i = 0; i < p->extra_family_id_count; i++, pos_++) {
+            p->extra_family_ids[i] = extra_family_ids_and_name[pos_];
+        }
+
+        if (p->has_name) {
+            uint8_t *name_buf = (uint8_t *)&extra_family_ids_and_name[pos_];
+            uint8_t name_length = *name_buf++ & 0x7F;
+            memcpy(p->name, name_buf, name_length);
+            p->name[name_length] = '\0';
+        }
+    }
+    if (!p->has_name)
+         p->name[0] = '\0';
+
+    pt->current_partition++;
+    return true;
+}
+
+int main() {
+    stdio_init_all();
+
+    pico_partition_table_t pt;
+    int rc;
+    rc = read_partition_table(&pt);
+    if (rc != 0) {
+        panic("rom_get_partition_table_info returned %d", pt.status);
+    }
+    if (!pt.has_partition_table) {
+        printf("there is no partition table\n");
+    } else if (pt.partition_count == 0) {
+        printf("the partition table is empty\n");
+    }
+
+    uf2_family_ids_t *family_ids = uf2_family_ids_new(pt.flags_and_permissions);
+    char *str_family_ids = uf2_family_ids_join(family_ids, ", ");
+    printf("un-partitioned_space: S(%s%s) NSBOOT(%s%s) NS(%s%s) uf2 { %s }\n",
+           (pt.flags_and_permissions & PICOBIN_PARTITION_PERMISSION_S_R_BITS ? "r" : ""),
+           (pt.flags_and_permissions & PICOBIN_PARTITION_PERMISSION_S_W_BITS ? "w" : ""),
+           (pt.flags_and_permissions & PICOBIN_PARTITION_PERMISSION_NSBOOT_R_BITS ? "r" : ""),
+           (pt.flags_and_permissions & PICOBIN_PARTITION_PERMISSION_NSBOOT_W_BITS ? "w" : ""),
+           (pt.flags_and_permissions & PICOBIN_PARTITION_PERMISSION_NS_R_BITS ? "r" : ""),
+           (pt.flags_and_permissions & PICOBIN_PARTITION_PERMISSION_NS_W_BITS ? "w" : ""),
```

Comment:
Similarly to how you have a function (`uf2_family_ids_join`) for creating the formatted family-ids string, I wonder if it might be worth pulling out a separate function for creating the `S(%s%s) NSBOOT(%s%s) NS(%s%s)` part of this string? (which I believe you could then also use in the per-partition loop?) :thinking:

Reply:
Haha, I was about to suggest that you could use `%c` instead of `%s` in your `printf` formatting, but then I realised that you can't do "empty string" when using `char` :joy: